File: api/bot.py
import os
import logging
import json
from dotenv import load_dotenv
from telegram import Update, InlineKeyboardButton, InlineKeyboardMarkup
from telegram.ext import (
    ApplicationBuilder, CommandHandler, CallbackQueryHandler, MessageHandler, filters
)
from fastapi import FastAPI, Request
import uvicorn

# Import Supabase helper functions from your lib folder
from lib.supabase import (
    is_user_registered, register_user, get_registered_user_count,
    get_registered_users_list, log_admin_action
)

# Load environment variables for local development
load_dotenv()

# --- Environment Variables ---
BOT_TOKEN = os.environ.get("BOT_TOKEN")
ADMIN_ID = int(os.environ.get("ADMIN_ID")) if os.environ.get("ADMIN_ID") else None
ADMIN_USERNAME = os.environ.get("ADMIN_USERNAME") # Optional

# Basic error checking for critical variables
if not BOT_TOKEN:
    logging.error("BOT_TOKEN not found. Bot cannot start.")
    exit(1) # In a serverless context, this exits the current invocation

if ADMIN_ID is None:
    logging.warning("ADMIN_ID not set. Admin commands will not function correctly.")

# --- Bot Setup ---
# Configure logging
logging.basicConfig(
    format='%(asctime)s - %(name)s - %(levelname)s - %(message)s', level=logging.INFO
)
logger = logging.getLogger(__name__)

# Create the Telegram Application
application = ApplicationBuilder().token(BOT_TOKEN).build()

# --- Middleware for Admin Check ---
async def is_admin(update: Update, context):
    if update.effective_user.id == ADMIN_ID:
        return True
    await update.message.reply_text("Unauthorized access. This command is for admins only.")
    return False

# --- /start Command ---
async def start_command(update: Update, context):
    user = update.effective_user
    telegram_id = user.id
    username = user.username
    first_name = user.first_name

    logger.info(f"[START COMMAND] User {telegram_id} ({username or 'N/A'}) initiated /start.")

    try:
        registered = await is_user_registered(telegram_id)

        if registered is None:
            logger.error(f"[START COMMAND] Supabase error during is_user_registered for user {telegram_id}.")
            await update.message.reply_text("Something went wrong, please try again later.")
            return

        if registered:
            logger.info(f"[START COMMAND] User {telegram_id} already registered.")
            await update.message.reply_text("You’re already registered. I’ll notify you when the app is live.")
        else:
            # User is NOT registered, prompt for confirmation
            registration_prompt = (
                f"Hello {first_name or 'there'}! I see you're not yet registered.\n\n"
                "I'll collect the following information to keep you updated:\n"
                f"• Your Telegram ID: `{telegram_id}`\n"
                f"• Your Username: `@{username}`\n" if username else "• Your Username: `Not available` (You can set one in Telegram settings!)\n"
                f"• Your First Name: `{first_name or 'Not provided'}`\n\n"
                "Would you like to register for updates about the Portfolio Showcase app?"
            )

            keyboard_buttons = [
                [
                    InlineKeyboardButton("✅ Yes, register me!", callback_data=f"register_yes:{telegram_id}"),
                    InlineKeyboardButton("❌ No, thanks.", callback_data="register_no")
                ]
            ]

            if ADMIN_USERNAME:
                keyboard_buttons.append([
                    InlineKeyboardButton("❓ Contact Admin", url=f"https://t.me/{ADMIN_USERNAME}")
                ])

            reply_markup = InlineKeyboardMarkup(keyboard_buttons)
            logger.info(f"[START COMMAND] Sending registration prompt with buttons to {telegram_id}.")
            await update.message.reply_text(
                registration_prompt,
                reply_markup=reply_markup,
                parse_mode='Markdown'
            )

    except Exception as e:
        logger.exception("[START COMMAND] Unhandled error in /start command.")
        await update.message.reply_text("Something went wrong, please try again later.")

# --- Callback Query Handler for Registration Buttons ---
async def registration_callback(update: Update, context):
    query = update.callback_query
    await query.answer() # Acknowledge the callback query

    callback_data = query.data
    action = callback_data.split(':')[0]
    telegram_id_from_callback = int(callback_data.split(':')[1]) if ':' in callback_data else None
    current_telegram_id = query.from_user.id
    first_name = query.from_user.first_name or 'there'

    logger.info(f"[CALLBACK ACTION] User {current_telegram_id} clicked {action}.")

    # Security check: ensure the button click is from the user who initiated it
    if telegram_id_from_callback and telegram_id_from_callback != current_telegram_id:
        await query.edit_message_text("This registration prompt is not for you.")
        return

    try:
        if action == 'register_yes':
            username = query.from_user.username
            
            already_registered = await is_user_registered(current_telegram_id)
            if already_registered is None:
                logger.error(f"[CALLBACK ACTION] Supabase error during is_user_registered for action 'yes'.")
                await query.edit_message_text("Something went wrong, please try again later.")
                return
            if already_registered:
                logger.info(f"[CALLBACK ACTION] User {current_telegram_id} already registered, clicked 'yes'.")
                await query.edit_message_text(
                    "You’re already registered. I’ll notify you when the app is live.",
                    parse_mode='Markdown'
                )
                return

            user_data = {
                "telegram_id": current_telegram_id,
                "username": username,
                "first_name": query.from_user.first_name or 'N/A'
            }

            registration_result = await register_user(user_data)

            if registration_result:
                logger.info(f"[CALLBACK ACTION] User {current_telegram_id} successfully registered.")
                await query.edit_message_text(
                    f"🎉 Great! Thanks for registering, {first_name}! I’ll notify you when the Portfolio Showcase app is ready.",
                    parse_mode='Markdown'
                )
            else:
                logger.error(f"[CALLBACK ACTION] Failed to register user {current_telegram_id}.")
                await query.edit_message_text(
                    "Something went wrong during registration, please try again later.",
                    parse_mode='Markdown'
                )
        elif action == 'register_no':
            logger.info(f"[CALLBACK ACTION] User {current_telegram_id} clicked 'no'.")
            await query.edit_message_text(
                "No problem! You can type /start again anytime if you change your mind.",
                parse_mode='Markdown'
            )
    except Exception as e:
        logger.exception("[CALLBACK ACTION] Unhandled error in registration action.")
        await query.edit_message_text("Something went wrong, please try again later.")

# --- /help Command ---
async def help_command(update: Update, context):
    user_id = update.effective_user.id
    logger.info(f"[HELP COMMAND] User {user_id} initiated /help.")

    message = "Welcome to the Portfolio Showcase Bot!\n\n"
    message += "Use /start to register for updates on the upcoming Portfolio Showcase Mini App.\n"

    if user_id == ADMIN_ID:
        message += "\n--- Admin Commands ---\n"
        message += "/count - Get the total number of registered users.\n"
        message += "/list - Get a list of all registered usernames and IDs.\n"
        message += "/notify <message> - Send a broadcast message to all registered users. Example: `/notify The app is live!`\n"

    await update.message.reply_text(message)

# --- Admin Commands ---
async def count_command(update: Update, context):
    if not await is_admin(update, context): return
    admin_id = update.effective_user.id
    logger.info(f"[ADMIN COMMAND] Admin {admin_id} initiated /count.")

    try:
        count = await get_registered_user_count()
        if count is None:
            logger.error("[ADMIN COMMAND] Supabase error during get_registered_user_count.")
            await update.message.reply_text("Something went wrong while fetching user count, please try again later.")
            await log_admin_action({"admin_telegram_id": admin_id, "action": "count_users_failed", "details": "Database error fetching count."})
            return

        await update.message.reply_text(f"Currently, {count} users are registered.")
        await log_admin_action({"admin_telegram_id": admin_id, "action": "count_users", "details": f"Returned count: {count}"})

    except Exception as e:
        logger.exception("[ADMIN COMMAND] Unhandled error in /count command.")
        await update.message.reply_text("Something went wrong, please try again later.")
        await log_admin_action({"admin_telegram_id": admin_id, "action": "count_users_failed", "details": f"Unhandled error: {str(e)}"})

async def list_command(update: Update, context):
    if not await is_admin(update, context): return
    admin_id = update.effective_user.id
    logger.info(f"[ADMIN COMMAND] Admin {admin_id} initiated /list.")

    try:
        users = await get_registered_users_list()
        if users is None:
            logger.error("[ADMIN COMMAND] Supabase error during get_registered_users_list.")
            await update.message.reply_text("Something went wrong while fetching the user list, please try again later.")
            await log_admin_action({"admin_telegram_id": admin_id, "action": "list_users_failed", "details": "Database error fetching list."})
            return

        if not users:
            logger.info("[ADMIN COMMAND] No registered users found for /list.")
            await update.message.reply_text("No users are currently registered.")
            await log_admin_action({"admin_telegram_id": admin_id, "action": "list_users", "details": "No registered users."})
            return

        user_list = "\n".join([
            f"- `{user['telegram_id']}`" + (f" (@{user['username']})" if user.get('username') else "")
            for user in users
        ])
        await update.message.reply_text(f"Registered Users:\n{user_list}", parse_mode='Markdown')
        await log_admin_action({"admin_telegram_id": admin_id, "action": "list_users", "details": f"Returned {len(users)} users."})

    except Exception as e:
        logger.exception("[ADMIN COMMAND] Unhandled error in /list command.")
        await update.message.reply_text("Something went wrong, please try again later.")
        await log_admin_action({"admin_telegram_id": admin_id, "action": "list_users_failed", "details": f"Unhandled error: {str(e)}"})

async def notify_command(update: Update, context):
    if not await is_admin(update, context): return
    admin_id = update.effective_user.id
    logger.info(f"[ADMIN COMMAND] Admin {admin_id} initiated /notify.")

    message_text = " ".join(context.args)
    if not message_text:
        logger.warning("[ADMIN COMMAND] Notify command used without message text.")
        await update.message.reply_text("Please provide a message to send. Example: `/notify The app is now live!`", parse_mode='Markdown')
        await log_admin_action({"admin_telegram_id": admin_id, "action": "notify_failed", "details": "No message text provided."})
        return

    try:
        users = await get_registered_users_list()
        if users is None:
            logger.error("[ADMIN COMMAND] Supabase error during get_registered_users_list for /notify.")
            await update.message.reply_text("Something went wrong while fetching the user list for notification, please try again later.")
            await log_admin_action({"admin_telegram_id": admin_id, "action": "notify_failed", "details": "Database error fetching user list for broadcast."})
            return

        if not users:
            logger.info("[ADMIN COMMAND] No registered users to notify.")
            await update.message.reply_text("No users are currently registered to notify.")
            await log_admin_action({"admin_telegram_id": admin_id, "action": "notify_attempt_no_users", "details": "No registered users to send broadcast to."})
            return

        successful_sends = 0
        failed_sends = 0
        failed_user_ids = []

        for user in users:
            try:
                await context.bot.send_message(chat_id=user['telegram_id'], text=message_text)
                successful_sends += 1
            except Exception as send_error:
                logger.warning(f"[ADMIN COMMAND] Failed to send message to user {user['telegram_id']} ({user.get('username', 'N/A')}): {send_error}")
                failed_sends += 1
                failed_user_ids.append(str(user['telegram_id']))

        await update.message.reply_text(f"Broadcast complete! Sent to {successful_sends} users. Failed for {failed_sends} users.")
        await log_admin_action({
            "admin_telegram_id": admin_id,
            "action": "broadcast_message",
            "details": f"Message: \"{message_text}\" | Sent: {successful_sends}, Failed: {failed_sends} (IDs: {', '.join(failed_user_ids) or 'N/A'})"
        })

    except Exception as e:
        logger.exception("[ADMIN COMMAND] Unhandled error in /notify command.")
        await update.message.reply_text("Something went wrong during the broadcast, please try again later.")
        await log_admin_action({"admin_telegram_id": admin_id, "action": "notify_failed", "details": f"Unhandled error during broadcast: {str(e)}"})

# --- Generic Message Handler ---
async def generic_message_handler(update: Update, context):
    if update.message and update.message.text and not update.message.text.startswith('/'):
        logger.info(f"[MESSAGE HANDLER] User {update.effective_user.id} sent non-command message: '{update.message.text}'.")
        await update.message.reply_text("I'm a registration bot! Please use commands like /start or /help.")

# --- Register Handlers ---
application.add_handler(CommandHandler("start", start_command))
application.add_handler(CommandHandler("help", help_command))
application.add_handler(CommandHandler("count", count_command))
application.add_handler(CommandHandler("list", list_command))
application.add_handler(CommandHandler("notify", notify_command))
application.add_handler(CallbackQueryHandler(registration_callback, pattern=r'register_(yes|no):\d+'))
application.add_handler(MessageHandler(filters.TEXT & ~filters.COMMAND, generic_message_handler))

# --- Vercel Serverless Function Export ---
# FastAPI app to handle Vercel's HTTP requests
app = FastAPI()

@app.post("/api/bot")
async def webhook_handler(request: Request):
    logger.info(f"[WEBHOOK] Received request: Method={request.method}, URL={request.url}")
    if request.method == "POST":
        # Get the update from the request body
        body = await request.json()
        update = Update.de_json(body, application.bot)
        logger.info("[WEBHOOK] Processing Telegram update...")

        # Process the update
        await application.process_update(update)
        logger.info("[WEBHOOK] Telegram update processed.")
        return {"status": "ok"}
    else:
        logger.info("[WEBHOOK] Non-POST request received. Responding with info.")
        return {"message": "This is the Telegram Bot webhook endpoint. Please send POST requests with Telegram updates."}

# This is for local testing with Uvicorn
# To run locally: uvicorn api.bot:app --reload --port 8000
if __name__ == "__main__":
    logger.info("Bot started in local polling mode (for direct testing). This block will NOT run on Vercel.")
    # For local long-polling, you would normally run:
    # application.run_polling()
    # For local webhook testing (if using FastAPI as a local server), you would run uvicorn
    uvicorn.run(app, host="0.0.0.0", port=8000)

File: api/bot.js
require('dotenv').config(); // Load environment variables from .env file (for local testing)

const { Telegraf, Markup } = require('telegraf'); // <-- Make sure Markup is imported here!
const {
    isUserRegistered,
    registerUser,
    getRegisteredUserCount,
    getRegisteredUsersList,
    logAdminAction
} = require('../lib/supabase'); // Import Supabase helper functions

// Environment variables
const BOT_TOKEN = process.env.BOT_TOKEN;
const ADMIN_ID = process.env.ADMIN_ID; // Stored as string, convert to number for comparison

// Check for BOT_TOKEN
if (!BOT_TOKEN) {
    console.error('Error: BOT_TOKEN not found in environment variables. Bot cannot start.');
    process.exit(1); // In a serverless function, this might just terminate the current invocation.
}

const bot = new Telegraf(BOT_TOKEN);

// --- Middleware for Admin Check ---
const isAdmin = (ctx, next) => {
    // Telegraf's ctx.from.id is a number, ADMIN_ID from env is a string, so convert ADMIN_ID
    if (ctx.from && ctx.from.id === parseInt(ADMIN_ID, 10)) {
        return next();
    }
    ctx.reply('Unauthorized access. This command is for admins only.');
};

// --- /start Command ---
bot.start(async (ctx) => {
    const telegramId = ctx.from.id;
    const username = ctx.from.username;
    const firstName = ctx.from.first_name;

    try {
        const registered = await isUserRegistered(telegramId);

        if (registered === null) {
            // Supabase error occurred
            return ctx.reply("Something went wrong, please try again later.");
        }

        if (registered) {
            ctx.reply("You’re already registered. I’ll notify you when the app is live.");
        } else {
            // User is NOT registered, prompt for confirmation
            let registrationPrompt = `Hello ${firstName || 'there'}! I see you're not yet registered.\n\n`;
            registrationPrompt += `I'll collect the following information to keep you updated:\n`;
            registrationPrompt += `• Your Telegram ID: \`${telegramId}\`\n`;
            if (username) {
                registrationPrompt += `• Your Username: \`@${username}\`\n`;
            } else {
                registrationPrompt += `• Your Username: \`Not available\` (You can set one in Telegram settings!)\n`;
            }
            registrationPrompt += `• Your First Name: \`${firstName || 'Not provided'}\`\n\n`;
            registrationPrompt += `Would you like to register for updates about the Portfolio Showcase app?`;

            await ctx.reply(registrationPrompt, {
                parse_mode: 'Markdown',
                reply_markup: Markup.inlineKeyboard([
                    Markup.button.callback('✅ Yes, register me!', `register_yes:${telegramId}`),
                    Markup.button.callback('❌ No, thanks.', 'register_no')
                ])
            });
        }
    } catch (error) {
        console.error('Error in /start command:', error);
        ctx.reply("Something went wrong, please try again later.");
    }
});

// --- Callback Query Handler for Registration Buttons ---
bot.action(/register_(yes|no):?(\d+)?/, async (ctx) => {
    const action = ctx.match[1]; // 'yes' or 'no'
    const telegramIdFromCallback = ctx.match[2] ? parseInt(ctx.match[2], 10) : null;
    const currentTelegramId = ctx.from.id; // The ID of the user who clicked the button

    // It's good practice to verify that the user clicking 'yes' is the same user who initiated /start
    // This prevents one user from registering another by clicking their button in a group, for example.
    if (telegramIdFromCallback && telegramIdFromCallback !== currentTelegramId) {
        await ctx.answerCbQuery('This registration prompt is not for you.', { show_alert: true });
        return;
    }

    // Dismiss the loading spinner on the button
    await ctx.answerCbQuery();

    try {
        if (action === 'yes') {
            const username = ctx.from.username;
            const firstName = ctx.from.first_name;

            // Double check registration status before trying to register
            const alreadyRegistered = await isUserRegistered(currentTelegramId);
            if (alreadyRegistered === null) {
                await ctx.editMessageText("Something went wrong, please try again later.");
                return;
            }
            if (alreadyRegistered) {
                await ctx.editMessageText(
                    "You’re already registered. I’ll notify you when the app is live.",
                    { parse_mode: 'Markdown' }
                );
                return;
            }

            const userData = {
                telegram_id: currentTelegramId,
                username: username,
                first_name: firstName || 'N/A'
            };

            const registrationResult = await registerUser(userData);

            if (registrationResult) {
                // Edit the original message to reflect the action
                await ctx.editMessageText(
                    `🎉 Great! Thanks for registering, ${firstName || 'there'}! I’ll notify you when the Portfolio Showcase app is ready.`,
                    { parse_mode: 'Markdown' }
                );
            } else {
                // This could happen if there's a race condition or actual DB error
                await ctx.editMessageText(
                    "Something went wrong during registration, please try again later.",
                    { parse_mode: 'Markdown' }
                );
            }
        } else if (action === 'no') {
            await ctx.editMessageText(
                "No problem! You can type /start again anytime if you change your mind.",
                { parse_mode: 'Markdown' }
            );
        }
    } catch (error) {
        console.error('Error in registration action:', error);
        await ctx.editMessageText("Something went wrong, please try again later.");
    }
});


// --- /help Command ---
bot.help(async (ctx) => {
    const adminId = parseInt(ADMIN_ID, 10);
    const userId = ctx.from.id;

    let message = "Welcome to the Portfolio Showcase Bot!\n\n";
    message += "Use /start to register for updates on the upcoming Portfolio Showcase Mini App.\n";

    if (userId === adminId) {
        message += "\n--- Admin Commands ---\n";
        message += "/count - Get the total number of registered users.\n";
        message += "/list - Get a list of all registered usernames and IDs.\n";
        message += "/notify <message> - Send a broadcast message to all registered users. Example: `/notify The app is live!`\n";
    }

    ctx.reply(message);
});

// --- Admin Commands ---

// /count command
bot.command('count', isAdmin, async (ctx) => {
    try {
        const count = await getRegisteredUserCount();

        if (count === null) {
            await ctx.reply("Something went wrong while fetching user count, please try again later.");
            await logAdminAction({
                admin_telegram_id: ctx.from.id,
                action: 'count_users_failed',
                details: 'Database error fetching count.'
            });
            return;
        }

        await ctx.reply(`Currently, ${count} users are registered.`);
        await logAdminAction({
            admin_telegram_id: ctx.from.id,
            action: 'count_users',
            details: `Returned count: ${count}`
        });

    } catch (error) {
        console.error('Error in /count command:', error);
        ctx.reply("Something went wrong, please try again later.");
        await logAdminAction({
            admin_telegram_id: ctx.from.id,
            action: 'count_users_failed',
            details: `Unhandled error: ${error.message}`
        });
    }
});

// /list command
bot.command('list', isAdmin, async (ctx) => {
    try {
        const users = await getRegisteredUsersList();

        if (users === null) {
            await ctx.reply("Something went wrong while fetching the user list, please try again later.");
            await logAdminAction({
                admin_telegram_id: ctx.from.id,
                action: 'list_users_failed',
                details: 'Database error fetching list.'
            });
            return;
        }

        if (users.length === 0) {
            await ctx.reply("No users are currently registered.");
            await logAdminAction({
                admin_telegram_id: ctx.from.id,
                action: 'list_users',
                details: 'No registered users.'
            });
            return;
        }

        // Format the list for display
        const userList = users.map(user => {
            const usernamePart = user.username ? ` (@${user.username})` : '';
            return `- ${user.telegram_id}${usernamePart}`;
        }).join('\n');

        await ctx.reply(`Registered Users:\n${userList}`, { parse_mode: 'Markdown' });
        await logAdminAction({
            admin_telegram_id: ctx.from.id,
            action: 'list_users',
            details: `Returned ${users.length} users.`
        });

    } catch (error) {
        console.error('Error in /list command:', error);
        ctx.reply("Something went wrong, please try again later.");
        await logAdminAction({
            admin_telegram_id: ctx.from.id,
            action: 'list_users_failed',
            details: `Unhandled error: ${error.message}`
        });
    }
});

// /notify <message> command
bot.command('notify', isAdmin, async (ctx) => {
    const messageText = ctx.message.text.substring('/notify '.length).trim();

    if (!messageText) {
        await ctx.reply("Please provide a message to send. Example: `/notify The app is now live!`", { parse_mode: 'Markdown' });
        await logAdminAction({
            admin_telegram_id: ctx.from.id,
            action: 'notify_failed',
            details: 'No message text provided.'
        });
        return;
    }

    try {
        const users = await getRegisteredUsersList();

        if (users === null) {
            await ctx.reply("Something went wrong while fetching the user list for notification, please try again later.");
            await logAdminAction({
                admin_telegram_id: ctx.from.id,
                action: 'notify_failed',
                details: 'Database error fetching user list for broadcast.'
            });
            return;
        }

        if (users.length === 0) {
            await ctx.reply("No users are currently registered to notify.");
            await logAdminAction({
                admin_telegram_id: ctx.from.id,
                action: 'notify_attempt_no_users',
                details: 'No registered users to send broadcast to.'
            });
            return;
        }

        let successfulSends = 0;
        let failedSends = 0;
        const failedUserIds = [];

        for (const user of users) {
            try {
                // Add a small delay to avoid hitting Telegram API rate limits if many users
                await new Promise(resolve => setTimeout(resolve, 50)); // 50ms delay
                await bot.telegram.sendMessage(user.telegram_id, messageText);
                successfulSends++;
            } catch (sendError) {
                console.warn(`Failed to send message to user ${user.telegram_id} (${user.username || 'N/A'}): ${sendError.message}`);
                failedSends++;
                failedUserIds.push(user.telegram_id);
            }
        }

        await ctx.reply(`Broadcast complete! Sent to ${successfulSends} users. Failed for ${failedSends} users.`);
        await logAdminAction({
            admin_telegram_id: ctx.from.id,
            action: 'broadcast_message',
            details: `Message: "${messageText}" | Sent: ${successfulSends}, Failed: ${failedSends} (IDs: ${failedUserIds.join(', ') || 'N/A'})`
        });

    } catch (error) {
        console.error('Error in /notify command:', error);
        ctx.reply("Something went wrong during the broadcast, please try again later.");
        await logAdminAction({
            admin_telegram_id: ctx.from.id,
            action: 'notify_failed',
            details: `Unhandled error during broadcast: ${error.message}`
        });
    }
});

// Generic error handler for Telegraf (optional, but good practice)
bot.catch((err, ctx) => {
    console.error(`[Telegraf Error] for ${ctx.updateType}`, err);
    // You might want to send a generic error message to the user here
    // ctx.reply('Oops, something went wrong!');
});

// Handle non-command messages (e.g., plain text replies).
// For this bot, we mostly expect commands, so we can give a default hint.
bot.on('message', async (ctx) => {
    if (!ctx.message.text.startsWith('/')) { // Only reply if it's not a command
        ctx.reply("I'm a registration bot! Please use commands like /start or /help.");
    }
});


// --- Vercel Serverless Function Export ---
// This is crucial for running on Vercel as a webhook
module.exports = async (req, res) => {
    try {
        // Ensure the request is a POST and has a body, typical for Telegram webhooks
        if (req.method === 'POST' && req.body) {
            await bot.handleUpdate(req.body);
        } else {
            // Respond to GET requests (e.g., browser access to the webhook URL)
            // or malformed POST requests.
            console.log('Received non-webhook request:', req.method, req.url);
            res.statusCode = 200; // Still respond 200, but perhaps with a different message
            res.setHeader('Content-Type', 'text/plain');
            res.end('Telegram Bot Webhook Endpoint. Please send POST requests with Telegram updates.');
            return;
        }
    } catch (err) {
        console.error('Error handling update:', err);
    } finally {
        // Vercel expects a response, even if the bot logic is async.
        // A 200 OK status indicates the webhook was received successfully.
        if (!res.headersSent) { // Ensure headers haven't already been sent by an early return
            res.statusCode = 200;
            res.setHeader('Content-Type', 'application/json');
            res.end(JSON.stringify({ success: true }));
        }
    }
};

// --- Local Development (Optional, for testing without Vercel) ---
// If you want to run this locally for testing without deploying to Vercel,
// you can uncomment the lines below and use bot.launch() with long polling.
// This block will NOT run on Vercel because Vercel only executes the module.exports function.
/*
if (process.env.NODE_ENV === 'development' && !process.env.VERCEL) {
    bot.launch();
    console.log('Bot started in long polling mode (local development)');
    // Enable graceful stop
    process.once('SIGINT', () => bot.stop('SIGINT'));
    process.once('SIGTERM', () => bot.stop('SIGTERM'));
}
*/